### PR TITLE
(OraklNode) Gopax support

### PR DESCRIPTION
**Pull request by nick-bisonai**
# Description

Support Gopax as one of data provider

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.

## Deployment

- [ ] Should publish npm package
- [x] Should publish Docker image

### CODE REVIEW

File: node/pkg/websocketfetcher/providers/gopax/gopax.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+		Subscription{
+			Name: "SubscribeToTickers",
+		},
+	}
+
+	ws, err := wss.NewWebsocketHelper(ctx,
+		wss.WithEndpoint(URL),
+		wss.WithSubscriptions(subscription),
+		wss.WithProxyUrl(config.Proxy),
+		wss.WithReadLimit(32769*10),
```
Comment (Intizar-T) [inline node/pkg/websocketfetcher/providers/gopax/gopax.go:37]: how about declaring these numbers as constants and leaving a small explanation (including the unit in the const name is also an option)

File: node/pkg/websocketfetcher/providers/gopax/gopax.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	if err != nil {
+		log.Error().Str("Player", "Gopax").Err(err).Msg("error in gopax.customReadFunc, failed to read from websocket")
+		return nil, err
+	}
+
+	rawResponse := string(data)
+
+	if strings.HasPrefix(rawResponse, "\"primus::ping::") {
+		log.Debug().Str("Player", "Gopax").Msg("received pong")
+		f.Ws.RawWrite(ctx, "\"primus::pong::"+rawResponse[15:])
```
Comment (Intizar-T) [inline node/pkg/websocketfetcher/providers/gopax/gopax.go:107]: I'm guessing adding `rawResponse[15:]` is necessary to keep the connection alive?

Comment (nick-bisonai) [reply node/pkg/websocketfetcher/providers/gopax/gopax.go:107]: yh I tried to follow the guideline written in their [official docs](https://gopax.github.io/wsapi/index.en.html?go#authentication)

File: node/pkg/websocketfetcher/providers/gopax/gopax.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	if response.Name == "TickerEvent" {
+		var tickers Tickers
+		err := json.Unmarshal(response.Object, &tickers)
+		if err != nil {
+			return fmt.Errorf("failed to parse response: %v", response)
+		}
+		for _, ticker := range tickers {
+			feedData, err := TickerToFeedData(ticker, f.FeedMap)
+			if err != nil {
+				if err.Error() == "feed not found" {
```
Comment (Intizar-T) [inline node/pkg/websocketfetcher/providers/gopax/gopax.go:77]: how about moving this error to a list of other errors, i think it was in `sentinel`?

Comment (nick-bisonai) [reply node/pkg/websocketfetcher/providers/gopax/gopax.go:77]: yh I was feeling bothered to do that 🥹
but yes, will do

File: node/pkg/websocketfetcher/providers/gopax/gopax.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+		Subscription{
+			Name: "SubscribeToTickers",
+		},
+	}
+
+	ws, err := wss.NewWebsocketHelper(ctx,
+		wss.WithEndpoint(URL),
+		wss.WithSubscriptions(subscription),
+		wss.WithProxyUrl(config.Proxy),
+		wss.WithReadLimit(327690),
```
Comment (Intizar-T) [inline node/pkg/websocketfetcher/providers/gopax/gopax.go:37]: how about replacing this value with `IncreasedReadLimit` from `gopax/type.go`? or does that const stand for something else?

Comment (nick-bisonai) [reply node/pkg/websocketfetcher/providers/gopax/gopax.go]: missed it 😂

**Review by Intizar-T — APPROVED** [review]
lgtm!

just out of curiosity, what's the read limit? i skimmed through their docs but couldnt find a reference. What would it mean for our app to have a read limit?

**Comment by nick-bisonai** [comment]
> lgtm!
> 
> just out of curiosity, what's the read limit? i skimmed through their docs but couldnt find a reference. What would it mean for our app to have a read limit?

gopax websocket returns information of all pair in single response for the initial response after subscription. this data is too long that app can't read it and returns an error without an increase of read limit. it is not written in their docs, but the default limit which was set by the websocket libarary which have been used were having less than their response. It was inevitable to update the setting

**Pull request merged**